### PR TITLE
Commit regiularly even if nothing was changed

### PR DESCRIPTION
**Pull request by gburanov**
Please see why do we need it here 

https://trello.com/c/gLmwWsoJ/2341-prevent-kafka-consumer-offsets-from-expiring-on-low-throughput-topics

PS: Tests are failing right now, will be fixed after @cedrics  commit his PR

### CODE REVIEW

**Review by akaspin — APPROVED** [review]
👍

File: config.go [hunk 1]
```diff
@@ -466,6 +469,8 @@ func (c *Config) Validate() error {
 		return ConfigurationError("Consumer.Retry.Backoff must be >= 0")
 	case c.Consumer.Offsets.CommitInterval <= 0:
 		return ConfigurationError("Consumer.Offsets.CommitInterval must be > 0")
+	case c.Consumer.Offsets.CommitForcedInterval <= 0:
+		return ConfigurationError("Consumer.Offsets.CommitEvenIfNotChangedInterval must be > 0")
```
Comment (itszootime) [inline config.go:473]: This should be `Consumer.Offsets.CommitForcedInterval`

File: config.go [hunk 2]
```diff
@@ -329,6 +331,7 @@ func NewConfig() *Config {
 	c.Consumer.MaxProcessingTime = 100 * time.Millisecond
 	c.Consumer.Return.Errors = false
 	c.Consumer.Offsets.CommitInterval = 1 * time.Second
+	c.Consumer.Offsets.CommitForcedInterval = 12 * time.Hour
```
Comment (itszootime) [inline config.go:334]: Let's set this to 1 hour, there's no harm in committing more often

File: config.go [hunk 3]
```diff
@@ -230,6 +230,8 @@ type Config struct {
 			// How frequently to commit updated offsets. Defaults to 1s.
 			CommitInterval time.Duration
 
+			CommitForcedInterval time.Duration
```
Comment (itszootime) [inline config.go:233]: Let's add some documentation to this, something like:

```
// How frequently to commit offsets, even when they have not been updated. Defaults to 1h.
```

File: offset_manager.go [hunk 4]
```diff
@@ -422,6 +424,7 @@ type brokerOffsetManager struct {
 	parent              *offsetManager
 	broker              *Broker
 	timer               *time.Ticker
+	commitForcedTimer   *time.Ticker
```
Comment (itszootime) [inline offset_manager.go:427]: `forcedTimer` would be more consistent with the `CommitInterval` timer

File: offset_manager.go [hunk 5]
```diff
@@ -446,13 +450,26 @@ func (om *offsetManager) newBrokerOffsetManager(broker *Broker) *brokerOffsetMan
 func (bom *brokerOffsetManager) mainLoop() {
 	for {
 		select {
+		case <-bom.commitForcedTimer.C:
+			if len(bom.subscriptions) > 0 {
+				for s := range bom.subscriptions {
+					if !s.offsetChangedSinceLastForced {
+						s.dirty = true
```
Comment (itszootime) [inline offset_manager.go:457]: This isn't thread safe

File: offset_manager.go [hunk 6]
```diff
@@ -446,13 +450,26 @@ func (om *offsetManager) newBrokerOffsetManager(broker *Broker) *brokerOffsetMan
 func (bom *brokerOffsetManager) mainLoop() {
 	for {
 		select {
+		case <-bom.commitForcedTimer.C:
+			if len(bom.subscriptions) > 0 {
```
Comment (itszootime) [inline offset_manager.go:454]: We should also check `bom.autoCommit` here too, if the commits are manual then we shouldn't interfere

File: offset_manager.go [hunk 7]
```diff
@@ -446,13 +450,26 @@ func (om *offsetManager) newBrokerOffsetManager(broker *Broker) *brokerOffsetMan
...
+				for s := range bom.subscriptions {
+					if !s.offsetChangedSinceLastForced {
```
Comment (itszootime) [inline offset_manager.go:456]: Do you think it's worth having this logic? I don't think an extra commit every hour will cause a problem

**Pull request merged**